File: src/main/java/kitchenpos/menus/tobe/menu/application/MenuServiceV2.java
package kitchenpos.menus.tobe.menu.application;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;
import kitchenpos.common.infra.Profanities;
import kitchenpos.common.tobe.domain.DisplayedName;
import kitchenpos.menus.tobe.menu.domain.model.Menu;
import kitchenpos.menus.tobe.menu.domain.model.MenuPrice;
import kitchenpos.menus.tobe.menu.domain.model.MenuProducts;
import kitchenpos.menus.tobe.menu.domain.repository.MenuRepository;
import kitchenpos.menus.tobe.menu.dto.MenuRequest;
import kitchenpos.menus.tobe.menu.infra.MenuProductsMapper;
import kitchenpos.menus.tobe.menugroup.domain.repository.MenuGroupRepositoryV2;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
public class MenuServiceV2 {

    private final MenuRepository menuRepository;
    private final MenuGroupRepositoryV2 menuGroupRepository;
    private final Profanities profanities;
    private final MenuProductsMapper menuProductsMapper;

    public MenuServiceV2(
        final MenuRepository menuRepository,
        final MenuGroupRepositoryV2 menuGroupRepository,
        final Profanities profanities,
        final MenuProductsMapper menuProductsMapper
    ) {
        this.menuRepository = menuRepository;
        this.menuGroupRepository = menuGroupRepository;
        this.profanities = profanities;
        this.menuProductsMapper = menuProductsMapper;
    }

    @Transactional
    public Menu create(final MenuRequest request) {
        menuGroupRepository.findById(request.getMenuGroupId())
            .orElseThrow(() -> new NoSuchElementException("해당 메뉴그룹을 찾을 수 없습니다."));
        final DisplayedName displayedName = new DisplayedName(request.getName(), profanities);
        final MenuPrice price = new MenuPrice(request.getPrice());
        final MenuProducts menuProducts = menuProductsMapper.toMenuProducts(request.getMenuProducts());
        final Menu menu = new Menu(displayedName, price, request.getMenuGroupId(), true, menuProducts);
        return menuRepository.save(menu);
    }

    @Transactional
    public Menu changePrice(final UUID menuId, final MenuPrice price) {
        final Menu menu = menuRepository.findById(menuId)
            .orElseThrow(() -> new NoSuchElementException("해당 메뉴를 찾을 수 없습니다."));
        return menu.changePrice(price);
    }

    @Transactional
    public Menu display(final UUID menuId) {
        final Menu menu = menuRepository.findById(menuId)
            .orElseThrow(() -> new NoSuchElementException("해당 메뉴를 찾을 수 없습니다."));
        return menu.display();
    }

    @Transactional
    public Menu hide(final UUID menuId) {
        final Menu menu = menuRepository.findById(menuId)
            .orElseThrow(() -> new NoSuchElementException("해당 메뉴를 찾을 수 없습니다."));
        return menu.hide();
    }

    @Transactional(readOnly = true)
    public List<Menu> findAll() {
        return menuRepository.findAll();
    }
}

File: src/main/kotlin/kitchenpos/tobe/menu/application/MenuServiceV2.kt
package kitchenpos.tobe.menu.application

import kitchenpos.tobe.menu.application.dto.request.ChangeMenuPriceRequest
import kitchenpos.tobe.menu.application.dto.request.CreateMenuRequest
import kitchenpos.tobe.menu.application.dto.response.ChangeMenuPriceResponse
import kitchenpos.tobe.menu.application.dto.response.CreateMenuResponse
import kitchenpos.tobe.menu.application.dto.response.GetMenusResponse
import kitchenpos.tobe.menu.domain.MenuPurgomalumClient
import kitchenpos.tobe.menu.domain.ProductClient
import kitchenpos.tobe.menu.domain.entity.MenuV2
import kitchenpos.tobe.menu.domain.repository.MenuGroupRepositoryV2
import kitchenpos.tobe.menu.domain.repository.MenuRepositoryV2
import kitchenpos.tobe.menu.domain.vo.MenuProductV2
import kitchenpos.tobe.menu.infra.getMenuById
import kitchenpos.tobe.menu.infra.getMenuGroupById
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class MenuServiceV2(
    private val menuGroupRepository: MenuGroupRepositoryV2,
    private val productClient: ProductClient,
    private val menuPurgomalumClient: MenuPurgomalumClient,
    private val menuRepository: MenuRepositoryV2,
) {
    fun createMenu(request: CreateMenuRequest): CreateMenuResponse {
        val menuGroup = menuGroupRepository.getMenuGroupById(request.menuGroupId)
        val quantityPerProductIds = request.menuProducts.associate { it.productId to it.quantity }
        val productPrices = productClient.getProductPrices(quantityPerProductIds.keys.toList())
        val menuProducts =
            productPrices.map {
                MenuProductV2.from(
                    productId = it.id,
                    price = it.price,
                    quantity = quantityPerProductIds[it.id]!!,
                )
            }.toMutableList()

        val menu =
            MenuV2.from(
                name = request.name,
                price = request.price,
                displayed = request.displayed,
                menuGroup = menuGroup,
                menuProducts = menuProducts,
                menuPurgomalumClient = menuPurgomalumClient,
            )

        val saved = menuRepository.save(menu)
        return CreateMenuResponse.of(saved)
    }

    fun changeMenuPrice(
        menuId: UUID,
        request: ChangeMenuPriceRequest,
    ): ChangeMenuPriceResponse {
        val menu = menuRepository.getMenuById(menuId)
        menu.changeMenuPrice(price = request.price)
        return ChangeMenuPriceResponse.of(menu)
    }

    fun display(menuId: UUID): UUID {
        val menu = menuRepository.getMenuById(menuId)
        menu.display()
        return menu.id
    }

    fun hide(menuId: UUID): UUID {
        val menu = menuRepository.getMenuById(menuId)
        menu.hide()
        return menu.id
    }

    fun findAll(): GetMenusResponse {
        val menus = menuRepository.findAll()
        return GetMenusResponse.of(menus)
    }
}
